### PR TITLE
Add PM usage counter to "device list"

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Useful for debugging, example (with https://github.com/zephyrproject-rtos/zephyr/pull/71788 as well):

```console
uart:~$ device list
devices:
- buttons (suspended, usage=0)
uart:~$ pm runtime-get buttons 
uart:~$ pm runtime-get buttons
uart:~$ device list
devices:
- buttons (active, usage=2)
uart:~$ pm runtime-put buttons
uart:~$ pm runtime-put buttons
uart:~$ device list
devices:
- buttons (suspended, usage=0)
```

### CODE REVIEW

**Comment by bjarki-trackunit** [comment]
Also big fan :)

**Comment by fabiobaltieri** [comment]
> I totally want this called out in the docs too :D This is awesome sauce!

should have seen that coming :-) let's wrap https://github.com/zephyrproject-rtos/zephyr/pull/71788 and then I can add a paragraph under "Device PM testing" that that PR introduce, moving this to draft in the meantime.

File: include/zephyr/pm/device_runtime.h [hunk 1]
```diff
@@ -158,6 +159,25 @@ int pm_device_runtime_put_async(const struct device *dev, k_timeout_t delay);
  */
 bool pm_device_runtime_is_enabled(const struct device *dev);
 
+/**
+ * @brief Return the current device usage counter.
+ *
+ * @param dev Device instance.
+ *
+ * @retval the current usage counter
```
Comment (gmarull) [inline include/zephyr/pm/device_runtime.h:167]:
```suggestion
 * @returns the current usage counter
```

File: subsys/shell/modules/device_service.c [hunk 2]
```diff
@@ -79,7 +79,17 @@ static int cmd_device_list(const struct shell *sh,
 #endif /* CONFIG_PM_DEVICE */
 		}
 
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+		if (dev->pm != NULL) {
+			uint32_t usage = pm_device_usage_get(dev);
```
Comment (gmarull) [inline subsys/shell/modules/device_service.c:84]: you should avoid accessing dev->pm, just provide a dummy impl for pm_device_usage_get in case PM is not built-in. Function could return int, <0 for error (-ENOSYS), >=0 for usage counter. It will also avoid ifdeffery in clients.

Comment (fabiobaltieri) [reply subsys/shell/modules/device_service.c:84]: wow, what was I thinking... there were other issues as well, namely returning a counter for device with normal pm, non-runtime, what's the opposite of "runtime" in this context? whatever, fixed, should handle all cases now, example:

```console
uart:~$ device list 
devices:
- clock@40000000 (READY)
- gpio@50000000 (READY)
- random@4000d000 (READY)
- entropy_bt_hci (READY)
- uart@40002000 (active)
- flash-controller@4001e000 (READY)
- i2c@40003000 (active)
- buttons (suspended, usage=0)
- leds (READY)
```

**Comment by fabiobaltieri** [comment]
Alright rebased and added a paragraph in the documentation.

File: include/zephyr/pm/device_runtime.h [hunk 3]
```diff
@@ -10,6 +10,7 @@
 
 #include <zephyr/device.h>
 #include <zephyr/kernel.h>
+#include <zephyr/pm/device.h>
```
Comment (gmarull) [inline include/zephyr/pm/device_runtime.h:13]: why is this needed?

File: include/zephyr/pm/device_runtime.h [hunk 4]
```diff
@@ -158,6 +159,17 @@ int pm_device_runtime_put_async(const struct device *dev, k_timeout_t delay);
...
+/**
+ * @brief Return the current device usage counter.
+ *
+ * @param dev Device instance.
+ *
+ * @returns the current usage counter.
+ * @retval -EINVAL If the device is not using runtime PM.
+ * @retval -ENOSYS If the runtime PM is not enabled at all.
+ */
+int pm_device_usage_get(const struct device *dev);
```
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply include/zephyr/pm/device_runtime.h]: yeah makes sense

File: include/zephyr/pm/device_runtime.h (hunk 3 above)
Comment (fabiobaltieri) [reply include/zephyr/pm/device_runtime.h]: good catch, it's a leftover for getting `struct pm_device` when this was an inline, but then I moved it to the normal file so this can go now

File: doc/services/pm/device.rst [hunk 5]
```diff
@@ -249,6 +249,22 @@ option and issue a ``pm`` command on a device from the shell, for example:
         devices:
         - buttons (suspended)
 
+To print the power management state of a device, enabled the
```
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply doc/services/pm/device.rst:252]: fixed

File: include/zephyr/pm/device_runtime.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply include/zephyr/pm/device_runtime.h]: ...or should it be just `pm_device_runtime_usage`? since there's already a `get` that does something else

[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply include/zephyr/pm/device_runtime.h]: yeah let me drop the `get` bit, it's just confusing

**Comment by fabiobaltieri** [comment]
- renamed the function to `pm_device_runtime_usage` (and did the other suggestions)

File: include/zephyr/pm/device_runtime.h [hunk 6]
```diff
@@ -158,6 +158,17 @@ int pm_device_runtime_put_async(const struct device *dev, k_timeout_t delay);
  */
 bool pm_device_runtime_is_enabled(const struct device *dev);
 
+/**
+ * @brief Return the current device usage counter.
+ *
+ * @param dev Device instance.
+ *
+ * @returns the current usage counter.
+ * @retval -EINVAL If the device is not using runtime PM.
```
Comment (gmarull) [inline include/zephyr/pm/device_runtime.h:167]:
```suggestion
 * @retval -ENOTSUP If the device is not using runtime PM.
```

**Review by gmarull — DISMISSED** [review]

**Review by gmarull — DISMISSED** [review]

**Review by tmleman — DISMISSED** [review]

**Review by ceolin — DISMISSED** [review]

**Review by henrikbrixandersen — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
Added tests for this one as well. Definitely not because I misread what PR @henrikbrixandersen asked me to add tests for.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
@ceolin can you do another review please? Same as what you've seen already but with tests.

**Review by JordanYates — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Comment by ceolin** [comment]
> @ceolin can you do another review please? Same as what you've seen already but with tests.

Thank you !
